pkg/kallsyms/kallsyms_test.go

```diff
@@ -77,78 +82,153 @@ func TestRealKAllSyms(t *testing.T) {
 	require.NotEqual(t, 0, addr, "bpf_prog_fops has a zero address")
 }
 
-func TestSpecRewriteConstants(t *testing.T) {
-	kAllSymsFactory := func() (*KAllSyms, error) {
-		kAllSymsStr := strings.Join([]string{
-			"0000000000000000 A fixed_percpu_data",
-			"ffffffffb4231f40 D bpf_prog_fops",
-			"ffffffffb43723e0 d socket_file_ops",
-		}, "\n")
-
-		kAllSymsReader := strings.NewReader(kAllSymsStr)
-		kAllSyms, err := NewKAllSymsFromReader(kAllSymsReader)
-		require.Nil(t, err, "NewKAllSymsFromReader failed: %v", err)
-		return kAllSyms, nil
-	}
-
-	// Little endian representation of the addresses above:
-	bpfProgFopsAddr := []byte{0x40, 0x1f, 0x23, 0xb4, 0xff, 0xff, 0xff, 0xff}
-	socketFileOpsAddr := []byte{0xe0, 0x23, 0x37, 0xb4, 0xff, 0xff, 0xff, 0xff}
-
-	spec := &ebpf.CollectionSpec{
-		Maps: map[string]*ebpf.MapSpec{
-			".rodata": {
-				Type:       ebpf.Array,
-				KeySize:    4,
-				ValueSize:  4,
-				MaxEntries: 1,
-				Value: &btf.Datasec{
-					Vars: []btf.VarSecinfo{
-						{
-							Type: &btf.Var{
-								Name: "bpf_prog_fops_addr",
-								Type: &btf.Int{Size: 8},
-							},
-							Offset: 0,
-							Size:   8,
+var specTest = &ebpf.CollectionSpec{
+	Maps: map[string]*ebpf.MapSpec{
+		".rodata": {
+			Type:       ebpf.Array,
+			KeySize:    4,
+			ValueSize:  4,
+			MaxEntries: 1,
+			Value: &btf.Datasec{
+				Vars: []btf.VarSecinfo{
+					{
+						Type: &btf.Var{
+							Name: "bpf_prog_fops_addr",
+							Type: &btf.Int{Size: 8},
 						},
-						{
-							Type: &btf.Var{
-								Name: "socket_file_ops_addr",
-								Type: &btf.Int{Size: 8},
-							},
-							Offset: 8,
-							Size:   8,
+						Offset: 0,
+						Size:   8,
+					},
+					{
+						Type: &btf.Var{
+							Name: "socket_file_ops_addr",
+							Type: &btf.Int{Size: 8},
 						},
+						Offset: 8,
+						Size:   8,
 					},
 				},
-				Contents: []ebpf.MapKV{
-					{Key: uint32(0), Value: []byte{
-						0, 0, 0, 0, 0, 0, 0, 0,
-						0, 0, 0, 0, 0, 0, 0, 0,
-					}},
-				},
+			},
+			Contents: []ebpf.MapKV{
+				{Key: uint32(0), Value: []byte{
+					0, 0, 0, 0, 0, 0, 0, 0,
+					0, 0, 0, 0, 0, 0, 0, 0,
+				}},
 			},
 		},
-	}
+	},
+}
+
+func TestSpecRewriteConstants(t *testing.T) {
+	t.Cleanup(resetState)
 
-	err := specUpdateAddresses(
-		kAllSymsFactory,
+	var err error
+
+	kAllSymsStr := strings.Join([]string{
+		"0000000000000000 A fixed_percpu_data",
+		"ffffffffb4231f40 D bpf_prog_fops",
+		"ffffffffb43723e0 d socket_file_ops",
+	},
+		"\n")
+	kAllSymsReader := strings.NewReader(kAllSymsStr)
+	kAllSyms, err := NewKAllSymsFromReader(kAllSymsReader)
+	require.NoError(t, err, "NewKAllSymsFromReader failed: %v", err)
+
+	// Little endian representation of the addresses above:
+	bpfProgFopsAddr := []byte{0x40, 0x1f, 0x23, 0xb4, 0xff, 0xff, 0xff, 0xff}
+	socketFileOpsAddr := []byte{0xe0, 0x23, 0x37, 0xb4, 0xff, 0xff, 0xff, 0xff}
```

Comment:
Good idea. Do you suggest to handle it in the same PR or in a separate PR? Also, regarding implementation I was wondering if we should have a separate package for [`NativeEndian`](https://github.com/inspektor-gadget/inspektor-gadget/blob/b2479d43f4de4e65ce16a17b76ee41f3cd046c70/pkg/datasource/endian_be.go) ?

Reply:
Since we use go 1.21 (see go.mod), we don't need to define NativeEndian ourselves anymore. We can use binary.NativeEndian:
https://pkg.go.dev/encoding/binary#pkg-variables

See [go 1.21 release notes](https://tip.golang.org/doc/go1.21):

> [encoding/binary](https://tip.golang.org/pkg/encoding/binary/)
> The new [NativeEndian](https://tip.golang.org/pkg/encoding/binary/#NativeEndian) variable may be used to convert between byte slices and integers using the current machine’s native endianness.

---

But since this PR didn't introduce the problem, please switch to NativeEndian in another PR once this one is merged.